### ISSUE TITLE
Support query batching

### ISSUE THREAD

**Issue by chillu** [silverstripe/silverstripe-graphql#177]
# Overview

Query batching allows for multiple queries to be processed in a single HTTP request, reducing network overheads, and allowing in-process optimisations. It is already part of the webonyx PHP library we're using (see http://webonyx.github.io/graphql-php/executing-queries/#query-batching). 

# Acceptance Criteria

 * A GraphQL client can submit [apollo style](https://dev-blog.apollodata.com/query-batching-in-apollo-63acfd859862) query batches
 * Query batching executes all queries in parallel (through async PHP?).
 * Query batching works with the upcoming persisted queries feature
 * Query batching is turned on for CMS requests

# Notes

 * Without async PHP, this feature will make overall UI performance *slower*
 * Assuming enough server capacity (workers, memory), this feature isn't as important with HTTP/2 (less network overhead)

**Comment by chillu** [comment]
Thanks for checking my logic there Bernie! Man, it's been half a year already on this, and no movement. So much to do :)

> arguably there is more efficiency in processing similar queries as one rather than spin up php for each one

But server efficiency != perceived user performance. It really depends on how many cores/threads/workers you have available and idle on the server.

>  If we allow query batching it allows client-side caching responses at a more granular level. In asset admin, for example, apollo would cache the individual file responses. This means that paging back and forth is more about getting IDs and the heavier fetching of file details only needs to be done on the first request or when something changes.

I don't see how query batching changes the apollo client cache characteristics. From my understanding, it just checks if it has run this query with the all the required fields and the same parameters. It doesn't matter if that individual query was fired off as a batch in the same HTTP request or not. See https://www.apollographql.com/docs/react/advanced/caching.html#readquery

**Comment by blueo** [comment]
haha yes a while ago now! The efficiency stuff seems fair there probably isn't much in it. 
On the cache - no it doens't change the client caching behaviour, but now your caching is more granular. If you say, delete a file, then the query for the page is different and all the items on the page are re-fetched rather than just the info for IDs that have changed. 

But this all aside probably my main desire for batching comes from wanting to co-locate the fetching/data with the components that use them. Doing a large query at the 'top' means having to store the fetched state somewhere like redux and wire that up or pass the query props down through a considerable number of components - this binds all the components together by needing to pass props and makes it hard to split things up. This us an extensibility problem as it makes it more difficult to replace parts of the app via injector etc.

**Comment by chillu** [comment]
> If you say, delete a file, then the query for the page is different and all the items on the page are re-fetched rather than just the info for IDs that have changed.

I'm not sure we're on the same page of what is considered "query batching". Here's two examples of batched queries:

Multiple queries returning a single record

```
[
  {
    "query": "{getFile(id: 1)} { id }"
  },
  {
    "query": "{getFile(id: 2)} { id }"
  }
]
```

Multiple queries returning multiple records

```
[
  {
    "query": "{getFiles()} { id }"
  },
  {
    "query": "{getPages(i)} { id }"
  }
]
```

I agree that the granularity of data you fetch is important (and how much you pass through component props), but that's separate from how you send those queries over the wire (batched as a single HTTP request, or with multiple HTTP requests). As far as I understand it, both batched and non-batched have exactly the same client-side caching behaviour.

**Comment by blueo** [comment]
I think I'm coming at it from how asset admin currently seems to work. I does a top level single ` query ReadFiles`. That fetches files and their info for the current page. default caching for this is at the query level so if the page/query changes - the whole query is re-fetched, including all the file details. 

I don't think this is separate from isolating components - having the above structure requires passing down the state via props. Where as batching (or individual requests as you point out) would allow individual caches on the smaller components. 

We did try using individual requests in asset admin (with http2) and it quickly became a performance issue - which was 'solved' by allowing us to extend the top level query mentioned above. So it would seem that individual queries does have more overhead to a query that fetches all at once. Batching would allow both scenarios - individual queries are sent but the server can deal with them as a group.

**Comment by chillu** [comment]
I think we're on the same page? There's two things going on here: How much data you fetch (and re-fetch) in a single query, and how many queries you put into a single HTTP request. GraphQL query batching is technically about the latter. It can help make the former less impactful on both client and server experience (particularly without HTTP/2).

**Comment by kinglozzer** [comment]
@unclecheese I have an increasingly apparent need for this functionality. Before I think about spending more time on it - do you think https://github.com/silverstripe/silverstripe-graphql/pull/429 is too invasive? Looking back at it now, it’s going to take a lot of effort to tidy up, review and document, and it doesn’t really accomplish much more than https://github.com/silverstripe/silverstripe-graphql/pull/418 did.

I’m wondering if I’ve overcomplicated this and should just close the more complex PR and resurrect the simpler one, as there aren’t any API or BC-breaking changes in that one. What do you think?

**Comment by kinglozzer** [comment]
I’ve decided to roll with the simpler solution as [a module](https://github.com/bigfork/silverstripe-graphql-batching) because we’re getting close to putting our project live and I don’t like running forks in production 😅. It could quite easily be folded into this module if there’s appetite for it, probably as opt-in behaviour given the small DDoS risk associated with it.
